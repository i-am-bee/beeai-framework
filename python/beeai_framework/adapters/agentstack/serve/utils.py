# Copyright 2025 Â© BeeAI a Series of LF Projects, LLC
# SPDX-License-Identifier: Apache-2.0

from collections.abc import Generator
from typing import Annotated

from beeai_framework.adapters.agentstack.serve.server import AgentStackMemoryManager
from beeai_framework.agents import AnyAgent
from beeai_framework.backend import AssistantMessage, MessageTextContent, MessageToolCallContent, ToolMessage
from beeai_framework.backend.message import AnyMessage
from beeai_framework.serve import MemoryManager, init_agent_memory

try:
    import agentstack_sdk.a2a.extensions as agentstack_extensions
    import agentstack_sdk.a2a.types as agentstack_types
    import agentstack_sdk.server.context as agentstack_context

    from beeai_framework.adapters.a2a.agents._utils import convert_a2a_to_framework_message
except ModuleNotFoundError as e:
    raise ModuleNotFoundError(
        "Optional module [agentstack not found.\nRun 'pip install \"beeai-framework[agentstack]\"' to install."
    ) from e


def send_message_trajectory(
    msg: AnyMessage,
    trajectory: Annotated[
        agentstack_extensions.TrajectoryExtensionServer, agentstack_extensions.TrajectoryExtensionSpec()
    ],
) -> Generator[agentstack_types.Metadata[str, agentstack_extensions.Trajectory]]:
    if isinstance(msg, AssistantMessage):
        for content in msg.content:
            if isinstance(content, MessageTextContent):
                yield trajectory.trajectory_metadata(title="assistant", content=content.text)
            elif isinstance(content, MessageToolCallContent):
                if content.tool_name == "final_answer":
                    continue
                yield trajectory.trajectory_metadata(title=f"{content.tool_name} (request)", content=content.args)
    elif isinstance(msg, ToolMessage):
        for tool_call in msg.get_tool_results():
            if tool_call.tool_name == "final_answer":
                continue

            yield trajectory.trajectory_metadata(
                title=f"{tool_call.tool_name} (response)", content=str(tool_call.result)
            )


async def init_agent_stack_memory(
    agent: AnyAgent, memory_manager: MemoryManager, context: agentstack_context.RunContext
) -> None:
    if isinstance(memory_manager, AgentStackMemoryManager):
        history = [message async for message in context.load_history() if message.parts]
        agent.memory.reset()
        await agent.memory.add_many([convert_a2a_to_framework_message(message) for message in history])
    else:
        await init_agent_memory(agent, memory_manager, context.context_id)
