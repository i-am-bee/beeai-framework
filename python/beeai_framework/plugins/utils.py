# Copyright 2025 Â© BeeAI a Series of LF Projects, LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import inspect
from collections.abc import Callable
from functools import cached_property
from typing import Any, ClassVar, TypeVar, overload

from pydantic import BaseModel
from typing_extensions import Unpack

from beeai_framework.context import Run, RunContext
from beeai_framework.emitter import Emitter
from beeai_framework.plugins.plugin import AnyPlugin, Plugin, PluginKwargs
from beeai_framework.utils.models import ModelLike, get_input_schema, get_output_schema, to_model
from beeai_framework.utils.strings import to_safe_word

TInput = TypeVar("TInput", bound=BaseModel)
TOutput = TypeVar("TOutput", bound=BaseModel)

TFunction = Callable[..., Any]


@overload
def plugin(
    plugin_function: TFunction,
    /,
    *,
    name: str | None = ...,
    description: str | None = ...,
    input_schema: type[TInput] | None = ...,
    output_schema: type[TOutput] | None = ...,
    emitter: Emitter | None = None,
) -> AnyPlugin: ...
@overload
def plugin(
    *,
    name: str | None = ...,
    description: str | None = ...,
    input_schema: type[TInput] | None = ...,
    output_schema: type[TOutput] | None = ...,
    emitter: Emitter | None = None,
) -> Callable[[TFunction], AnyPlugin]: ...
def plugin(
    fn: TFunction | None = None,
    /,
    *,
    name: str | None = None,
    description: str | None = None,
    input_schema: type[TInput] | None = None,
    output_schema: type[TOutput] | None = None,
    emitter: Emitter | None = None,
) -> AnyPlugin | Callable[[TFunction], AnyPlugin]:
    def create_plugin(fn: TFunction) -> AnyPlugin:
        plugin_name = name or fn.__name__
        plugin_description = description or inspect.getdoc(fn)
        plugin_input: type[TInput] = input_schema or get_input_schema(fn)
        plugin_output: type[TOutput] = output_schema or get_output_schema(fn)

        if plugin_description is None:
            raise ValueError("No plugin description provided.")

        class FunctionPlugin(Plugin[TInput, TOutput]):
            _auto_register: ClassVar[bool] = False

            name = plugin_name
            description = plugin_description or ""
            input_schema = plugin_input
            output_schema = plugin_output

            @cached_property
            def emitter(self) -> Emitter:
                if emitter is not None:
                    return emitter

                return Emitter.root().child(
                    namespace=["plugin", "custom", to_safe_word(self.name)],
                    creator=self,
                )

            def run(self, input: ModelLike[TInput], /, **kwargs: Unpack[PluginKwargs]) -> Run[TOutput]:
                async def handler(context: RunContext) -> TOutput:
                    target_input = to_model(self.input_schema, input).model_dump()
                    if inspect.iscoroutinefunction(fn):
                        result = await fn(**target_input)
                    else:
                        result = fn(**target_input)

                    return to_model(plugin_output, result)

                return RunContext.enter(self, handler, signal=kwargs.get("signal"), run_params={"input": input})

        return FunctionPlugin()

    if fn is None:
        return create_plugin
    else:
        return create_plugin(fn)
