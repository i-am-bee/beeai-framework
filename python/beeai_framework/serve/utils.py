# Copyright 2025 Â© BeeAI a Series of LF Projects, LLC
# SPDX-License-Identifier: Apache-2.0

from beeai_framework.agents import AnyAgent
from beeai_framework.logger import Logger
from beeai_framework.memory import BaseMemory
from beeai_framework.serve.memory_manager import MemoryManager

logger = Logger(__name__)


async def init_agent_memory(
    agent: AnyAgent, memory_manager: MemoryManager, session_id: str | None, *, stateful: bool = True
) -> None:
    async def create_empty_memory() -> BaseMemory:
        memory = await agent.memory.clone()
        memory.reset()
        return memory

    if stateful and session_id:
        if not await memory_manager.contains(session_id):
            await memory_manager.set(session_id, await create_empty_memory())
        memory = await memory_manager.get(session_id)
    else:
        memory = await create_empty_memory()

    try:
        agent.memory = memory
    except Exception:
        logger.debug("Agent does not support setting a new memory, resetting existing one for the agent.")
        agent.memory.reset()
