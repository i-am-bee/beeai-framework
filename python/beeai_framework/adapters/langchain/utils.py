# Copyright 2025 Â© BeeAI a Series of LF Projects, LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

try:
    from langchain_core.documents import Document
except ModuleNotFoundError as e:
    raise ModuleNotFoundError(
        "Optional module [langchain] not found.\nRun 'pip install \"beeai-framework[langchain]\"' to install."
    ) from e


from beeai_framework.adapters.langchain.wrappers.lc_embedding import LCEmbedding
from beeai_framework.backend.embedding import EmbeddingModel
from beeai_framework.backend.types import Document as VectorStoreDocument


def get_langchain_embedding(embedding_model: EmbeddingModel) -> LCEmbedding:
    return LCEmbedding(embedding=embedding_model)


def lc_document_to_document(lc_document: Document) -> VectorStoreDocument:
    return VectorStoreDocument(content=lc_document.page_content, metadata=lc_document.metadata)


def document_to_lc_document(document: VectorStoreDocument) -> Document:
    return Document(page_content=document.content, metadata=document.metadata)
