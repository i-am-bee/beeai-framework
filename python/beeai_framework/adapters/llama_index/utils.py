# Copyright 2025 Â© BeeAI a Series of LF Projects, LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

try:
    from llama_index.core.schema import NodeWithScore, TextNode
except ModuleNotFoundError as e:
    raise ModuleNotFoundError(
        "Optional module [llama_index] not found.\nRun 'pip install \"beeai-framework[llama_index]\"' to install."
    ) from e


from beeai_framework.backend.types import Document, DocumentWithScore


def doc_with_score_to_li_doc_with_score(document: DocumentWithScore) -> NodeWithScore:
    return NodeWithScore(
        node=TextNode(text=document.document.content, metadata=document.document.metadata), score=document.score
    )


def li_doc_with_score_to_doc_with_score(document: NodeWithScore) -> DocumentWithScore:
    return DocumentWithScore(document=Document(content=document.text, metadata=document.metadata), score=document.score)
