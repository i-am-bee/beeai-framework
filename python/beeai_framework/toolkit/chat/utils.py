# Copyright 2025 Â© BeeAI a Series of LF Projects, LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Util functions for the toolkit chat module."""

from typing import Any

from pydantic import BaseModel

from beeai_framework.backend.message import AnyMessage, UserMessage
from beeai_framework.plugins.schemas import DataContext
from beeai_framework.template import PromptTemplate
from beeai_framework.toolkit.chat.constants import USER_PROMPT_VARIABLE
from beeai_framework.toolkit.chat.template import FewShotChatPromptTemplate


def get_formatted_prompt(
    data: str | BaseModel | dict[str, Any] | UserMessage, template: PromptTemplate | FewShotChatPromptTemplate
) -> UserMessage:
    """Returns a prompt as a UserMessage.
    Args:
        data: the data to be formatted as a UserMessage.
        template: the template object to format the input.
    Returns:
        A formatted user message.
    """
    datum: dict[str, Any] = {}
    if isinstance(data, str):
        datum = {USER_PROMPT_VARIABLE: data}
    elif isinstance(data, dict):
        datum = data
    elif isinstance(data, BaseModel):
        datum = data.model_dump()
    elif isinstance(data, UserMessage):
        datum = {USER_PROMPT_VARIABLE: data.text}
    query = template.render(datum)
    return UserMessage(content=query)


def get_input_messages(cxt: DataContext, template: PromptTemplate | FewShotChatPromptTemplate) -> list[AnyMessage]:
    """Returns a prompt as a set of formatted messages.
    Args:
        cxt: the data to be formatted as a UserMessage.
        template: the template object to format the input.
    Returns:
        A list of formatted user messages.
    """
    input_messages: list[AnyMessage] = []
    data = cxt.data
    if isinstance(data, list) and len(data) > 0:
        for msg in data:
            if isinstance(msg, UserMessage):
                message = get_formatted_prompt(msg, template)
                input_messages.append(message)
            else:
                input_messages.append(msg)
    elif isinstance(data, str | dict | BaseModel | UserMessage):
        message = get_formatted_prompt(data, template)
        input_messages.append(message)
    return input_messages
