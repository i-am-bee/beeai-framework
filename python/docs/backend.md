# ⚙️ Backend

<!-- TOC -->
## Table of Contents
- [Overview](#overview)
- [Supported Providers](#supported-providers) 
- [Backend Initialization](#backend-initialization)
- [Chat Model](#chat-model)
  - [Chat Model Configuration](#chat-model-configuration)
  - [Text Generation](#text-generation)
  - [Streaming Responses](#streaming-responses)
  - [Structured Generation](#structured-generation)
  - [Tool Calling](#tool-calling)
- [Embedding Model](#embedding-model)
  - [Embedding Model Configuration](#embedding-model-configuration)
  - [Embedding Model Usage](#embedding-model-usage)
- [Troubleshooting](#troubleshooting)
- [Examples](#examples)
<!-- /TOC -->

---

## Overview

Backend is an umbrella module that encapsulates a unified way to work with the following functionalities:

- Chat Models via (`ChatModel` class)
- Embedding Models (coming soon)
- Audio Models (coming soon)
- Image Models (coming soon)

BeeAI framework's backend is designed with a provider-based architecture, allowing you to switch between different AI service providers while maintaining a consistent API.

> [!NOTE]
>
> Location within the framework: [beeai_framework/backend](/python/beeai_framework/backend).

--- 

## Supported providers

The following table depicts supported providers. Each provider requires specific configuration through environment variables. Ensure all required variables are set before initializing a provider.

| Name             | Chat | Embedding | Dependency               | Environment Variables                                                                                                                                                 |
| ---------------- | :--: | :-------: | ------------------------ | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `Ollama`         |  ✅  |          | `ollama-ai-provider`     | OLLAMA_CHAT_MODEL<br/>OLLAMA_BASE_URL                                                                                                       |
| `OpenAI`         |  ✅  |          | `openai`     | OPENAI_CHAT_MODEL<br/>OPENAI_API_BASE<br/>OPENAI_API_KEY<br/>OPENAI_ORGANIZATION                                                                                                       |
| `Watsonx`        |  ✅  |          | `@ibm-cloud/watsonx-ai`  | WATSONX_CHAT_MODEL<br/>WATSONX_EMBEDDING_MODEL<br>WATSONX_API_KEY<br/>WATSONX_PROJECT_ID<br/>WATSONX_SPACE_ID<br>WATSONX_VERSION<br>WATSONX_REGION                    |
| `Groq`           |    |         | Coming soon! | GROQ_CHAT_MODEL<br>GROQ_EMBEDDING_MODEL<br>GROQ_API_HOST<br>GROQ_API_KEY |
| `Amazon Bedrock` |    |         | Coming soon! | AWS_CHAT_MODEL<br>AWS_EMBEDDING_MODEL<br>AWS_ACCESS_KEY_ID<br>AWS_SECRET_ACCESS_KEY<br>AWS_REGION<br>AWS_SESSION_TOKEN |
| `Google Vertex`  |    |         | Coming soon! | GOOGLE_VERTEX_CHAT_MODEL<br>GOOGLE_VERTEX_EMBEDDING_MODEL<br>GOOGLE_VERTEX_PROJECT<br>GOOGLE_VERTEX_ENDPOINT<br>GOOGLE_VERTEX_LOCATION |
| `Azure OpenAI`   |    |         | Coming soon! | AZURE_OPENAI_CHAT_MODEL<br>AZURE_OPENAI_EMBEDDING_MODEL<br>AZURE_OPENAI_API_KEY<br>AZURE_OPENAI_API_ENDPOINT<br>AZURE_OPENAI_API_RESOURCE<br>AZURE_OPENAI_API_VERSION |
| `Anthropic`      |    |         | Coming soon! | ANTHROPIC_CHAT_MODEL<br>ANTHROPIC_EMBEDDING_MODEL<br>ANTHROPIC_API_KEY<br>ANTHROPIC_API_BASE_URL<br>ANTHROPIC_API_HEADERS |

> [!TIP]
>
> If you don't see your provider raise an issue [here](https://github.com/i-am-bee/beeai-framework/discussions). Meanwhile, you can use [Ollama adapter](/python/examples/backend/providers/ollama.py).

---

### Backend initialization

The `Backend` class serves as a central entry point to access models from your chosen provider.

<!-- embedme examples/backend/providers/watsonx.py -->

```py
import asyncio

from pydantic import BaseModel, Field

from beeai_framework.adapters.watsonx.backend.chat import WatsonxChatModel
from beeai_framework.backend.chat import ChatModel
from beeai_framework.backend.message import UserMessage
from beeai_framework.cancellation import AbortSignal
from beeai_framework.errors import AbortError

# Setting can be passed here during initiation or pre-configured via environment variables
llm = WatsonxChatModel(
    "ibm/granite-3-8b-instruct",
    # settings={
    #     "project_id": "WATSONX_PROJECT_ID",
    #     "api_key": "WATSONX_API_KEY",
    #     "api_base": "WATSONX_API_URL",
    # },
)


async def watsonx_from_name() -> None:
    watsonx_llm = ChatModel.from_name(
        "watsonx:ibm/granite-3-8b-instruct",
        # {
        #     "project_id": "WATSONX_PROJECT_ID",
        #     "api_key": "WATSONX_API_KEY",
        #     "api_base": "WATSONX_API_URL",
        # },
    )
    user_message = UserMessage("what states are part of New England?")
    response = await watsonx_llm.create(messages=[user_message])
    print(response.get_text_content())


async def watsonx_sync() -> None:
    user_message = UserMessage("what is the capital of Massachusetts?")
    response = await llm.create(messages=[user_message])
    print(response.get_text_content())


async def watsonx_stream() -> None:
    user_message = UserMessage("How many islands make up the country of Cape Verde?")
    response = await llm.create(messages=[user_message], stream=True)
    print(response.get_text_content())


async def watsonx_stream_abort() -> None:
    user_message = UserMessage("What is the smallest of the Cape Verde islands?")

    try:
        response = await llm.create(messages=[user_message], stream=True, abort_signal=AbortSignal.timeout(0.5))

        if response is not None:
            print(response.get_text_content())
        else:
            print("No response returned.")
    except AbortError as err:
        print(f"Aborted: {err}")


async def watson_structure() -> None:
    class TestSchema(BaseModel):
        answer: str = Field(description="your final answer")

    user_message = UserMessage("How many islands make up the country of Cape Verde?")
    response = await llm.create_structure(schema=TestSchema, messages=[user_message])
    print(response.object)


async def main() -> None:
    print("*" * 10, "watsonx_from_name")
    await watsonx_from_name()
    print("*" * 10, "watsonx_sync")
    await watsonx_sync()
    print("*" * 10, "watsonx_stream")
    await watsonx_stream()
    print("*" * 10, "watsonx_stream_abort")
    await watsonx_stream_abort()
    print("*" * 10, "watson_structure")
    await watson_structure()


if __name__ == "__main__":
    asyncio.run(main())

```

_Source: [examples/backend/providers/watsonx.py](/python/examples/backend/providers/watsonx.py)_

All providers examples can be found in [examples/backend/providers](/examples/backend/providers).

---

## Chat model

The `ChatModel` class represents a Chat Large Language Model and provides methods for text generation, streaming responses, and more. You can initialize a chat model in multiple ways:

**Method 1: Using the generic factory method**

```python
from beeai_framework.backend.chat import ChatModel

ollama_chat_model = ChatModel.from_name("ollama:llama3.1")
```

**Method 2: Creating a specific provider model directly**

```python
from beeai_framework.adapters.ollama.backend.chat import OllamaChatModel

ollama_chat_model = OllamaChatModel("llama3.1")
```

### Chat model configuration

You can configure various parameters for your chat model:

```txt
Coming soon
```

### Text generation

The most basic usage is to generate text responses:

```python
from beeai_framework.adapters.ollama.backend.chat import OllamaChatModel
from beeai_framework.backend.message import UserMessage

ollama_chat_model = OllamaChatModel("llama3.1")
response = await ollama_chat_model.create(
    messages=[UserMessage("what states are part of New England?")]
)

print(response.get_text_content())
```

> [!NOTE]
>
> Execution parameters (those passed to `model.create({...})`) are superior to ones defined via `config`.

### Streaming responses

For applications requiring real-time responses:

```python
from beeai_framework.adapters.ollama.backend.chat import OllamaChatModel
from beeai_framework.backend.message import UserMessage

llm = OllamaChatModel("llama3.1")
user_message = UserMessage("How many islands make up the country of Cape Verde?")
response = await llm.create(messages=[user_message], stream=True)
```

### Structured generation

Generate structured data according to a schema:

```txt
Coming soon
```

_Source: /examples/backend/structured.py_

### Tool calling

Integrate external tools with your AI model:

```txt
Coming soon
```

_Source: /examples/backend/toolCalling.py_

---

## Embedding model

The `EmbedingModel` class provides functionality for generating vector embeddings from text.

### Embedding model initialization

You can initialize an embedding model in multiple ways:

**Method 1: Using the generic factory method**

```txt
Coming soon
```

**Method 2: Creating a specific provider model directly**

```txt
Coming soon
```

### Embedding model usage

Generate embeddings for one or more text strings:

```txt
Coming soon
```

---

## Advanced usage

If your preferred provider isn't directly supported, you can use the LangChain adapter as a bridge. 

This allows you to leverage any provider that has LangChain compatibility.

```txt
Coming soon
```

_Source: /examples/backend/providers/langchain.py_

---

## Troubleshooting

Common issues and their solutions:

1. Authentication errors: Ensure all required environment variables are set correctly
2. Model not found: Verify that the model ID is correct and available for the selected provider

---

## Examples

- All backend examples can be found in [here](/python/examples/backend).