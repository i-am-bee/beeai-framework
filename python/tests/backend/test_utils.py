# Copyright 2025 Â© BeeAI a Series of LF Projects, LLC
# SPDX-License-Identifier: Apache-2.0

import pytest
from syrupy.assertion import SnapshotAssertion

from beeai_framework.backend.utils import generate_tool_union_schema
from beeai_framework.tools import AnyTool, tool


@tool()
def tool_sum(a: int, b: int = 0) -> int:
    """Sum tool"""

    return a + b


@tool()
def tool_greet(name: int) -> str:
    """Greet tool"""

    return f"Hello {name}!"


@pytest.mark.unit
@pytest.mark.parametrize("strict", [True, False])
def test_generate_tool_union_schema(strict: bool, snapshot: SnapshotAssertion) -> None:
    tools: list[AnyTool] = [tool_sum, tool_greet]
    result, _ = generate_tool_union_schema(
        tools, strict=strict, allow_parallel_tool_calls=False, allow_top_level_union=True
    )

    assert result["json_schema"] == snapshot()


@pytest.mark.unit
@pytest.mark.parametrize("strict", [True, False])
def test_generate_tool_union_schema_non_strict(strict: bool, snapshot: SnapshotAssertion) -> None:
    result, _ = generate_tool_union_schema(
        [tool_sum], strict=strict, allow_parallel_tool_calls=False, allow_top_level_union=True
    )
    assert result["json_schema"] == snapshot()


@pytest.mark.unit
def test_generate_tool_union_wrapped_schema(snapshot: SnapshotAssertion) -> None:
    tools: list[AnyTool] = [tool_sum, tool_greet]
    result, _ = generate_tool_union_schema(
        tools, strict=False, allow_parallel_tool_calls=True, allow_top_level_union=False
    )
    assert result["json_schema"] == snapshot()
