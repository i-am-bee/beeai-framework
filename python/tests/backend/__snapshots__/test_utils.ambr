# serializer version: 1
# name: test_generate_tool_union_schema[False]
  dict({
    'name': 'ToolCall',
    'schema': dict({
      'anyOf': list([
        dict({
          'additionalProperties': False,
          'properties': dict({
            'name': dict({
              'const': 'tool_sum',
              'description': 'Tool Name',
              'title': 'Name',
              'type': 'string',
            }),
            'parameters': dict({
              'additionalProperties': True,
              'description': 'Tool Parameters',
              'properties': dict({
                'a': dict({
                  'title': 'A',
                  'type': 'integer',
                }),
                'b': dict({
                  'default': 0,
                  'title': 'B',
                  'type': 'integer',
                }),
              }),
              'required': list([
                'a',
              ]),
              'title': 'tool_sum',
              'type': 'object',
            }),
          }),
          'required': list([
            'name',
            'parameters',
          ]),
          'title': 'tool_sum',
          'type': 'object',
        }),
        dict({
          'additionalProperties': False,
          'properties': dict({
            'name': dict({
              'const': 'tool_greet',
              'description': 'Tool Name',
              'title': 'Name',
              'type': 'string',
            }),
            'parameters': dict({
              'additionalProperties': True,
              'description': 'Tool Parameters',
              'properties': dict({
                'name': dict({
                  'title': 'Name',
                  'type': 'integer',
                }),
              }),
              'required': list([
                'name',
              ]),
              'title': 'tool_greet',
              'type': 'object',
            }),
          }),
          'required': list([
            'name',
            'parameters',
          ]),
          'title': 'tool_greet',
          'type': 'object',
        }),
      ]),
      'title': 'AvailableTools',
    }),
  })
# ---
# name: test_generate_tool_union_schema[True]
  dict({
    'name': 'ToolCall',
    'schema': dict({
      'anyOf': list([
        dict({
          'additionalProperties': False,
          'properties': dict({
            'name': dict({
              'const': 'tool_sum',
              'description': 'Tool Name',
              'title': 'Name',
              'type': 'string',
            }),
            'parameters': dict({
              'additionalProperties': True,
              'description': 'Tool Parameters',
              'properties': dict({
                'a': dict({
                  'title': 'A',
                  'type': 'integer',
                }),
                'b': dict({
                  'default': 0,
                  'title': 'B',
                  'type': 'integer',
                }),
              }),
              'required': list([
                'a',
                'b',
              ]),
              'title': 'tool_sum',
              'type': 'object',
            }),
          }),
          'required': list([
            'name',
            'parameters',
          ]),
          'title': 'tool_sum',
          'type': 'object',
        }),
        dict({
          'additionalProperties': False,
          'properties': dict({
            'name': dict({
              'const': 'tool_greet',
              'description': 'Tool Name',
              'title': 'Name',
              'type': 'string',
            }),
            'parameters': dict({
              'additionalProperties': True,
              'description': 'Tool Parameters',
              'properties': dict({
                'name': dict({
                  'title': 'Name',
                  'type': 'integer',
                }),
              }),
              'required': list([
                'name',
              ]),
              'title': 'tool_greet',
              'type': 'object',
            }),
          }),
          'required': list([
            'name',
            'parameters',
          ]),
          'title': 'tool_greet',
          'type': 'object',
        }),
      ]),
      'title': 'AvailableTools',
    }),
  })
# ---
# name: test_generate_tool_union_schema_non_strict[False]
  dict({
    'name': 'ToolCall',
    'schema': dict({
      'additionalProperties': False,
      'properties': dict({
        'name': dict({
          'const': 'tool_sum',
          'description': 'Tool Name',
          'title': 'Name',
          'type': 'string',
        }),
        'parameters': dict({
          'additionalProperties': True,
          'description': 'Tool Parameters',
          'properties': dict({
            'a': dict({
              'title': 'A',
              'type': 'integer',
            }),
            'b': dict({
              'default': 0,
              'title': 'B',
              'type': 'integer',
            }),
          }),
          'required': list([
            'a',
          ]),
          'title': 'tool_sum',
          'type': 'object',
        }),
      }),
      'required': list([
        'name',
        'parameters',
      ]),
      'title': 'tool_sum',
      'type': 'object',
    }),
  })
# ---
# name: test_generate_tool_union_schema_non_strict[True]
  dict({
    'name': 'ToolCall',
    'schema': dict({
      'additionalProperties': False,
      'properties': dict({
        'name': dict({
          'const': 'tool_sum',
          'description': 'Tool Name',
          'title': 'Name',
          'type': 'string',
        }),
        'parameters': dict({
          'additionalProperties': True,
          'description': 'Tool Parameters',
          'properties': dict({
            'a': dict({
              'title': 'A',
              'type': 'integer',
            }),
            'b': dict({
              'default': 0,
              'title': 'B',
              'type': 'integer',
            }),
          }),
          'required': list([
            'a',
            'b',
          ]),
          'title': 'tool_sum',
          'type': 'object',
        }),
      }),
      'required': list([
        'name',
        'parameters',
      ]),
      'title': 'tool_sum',
      'type': 'object',
    }),
  })
# ---
# name: test_generate_tool_union_wrapped_schema
  dict({
    'name': 'ToolCall',
    'schema': dict({
      'properties': dict({
        'item': dict({
          'items': dict({
            'anyOf': list([
              dict({
                'additionalProperties': False,
                'properties': dict({
                  'name': dict({
                    'const': 'tool_sum',
                    'description': 'Tool Name',
                    'title': 'Name',
                    'type': 'string',
                  }),
                  'parameters': dict({
                    'additionalProperties': True,
                    'description': 'Tool Parameters',
                    'properties': dict({
                      'a': dict({
                        'title': 'A',
                        'type': 'integer',
                      }),
                      'b': dict({
                        'default': 0,
                        'title': 'B',
                        'type': 'integer',
                      }),
                    }),
                    'required': list([
                      'a',
                    ]),
                    'title': 'tool_sum',
                    'type': 'object',
                  }),
                }),
                'required': list([
                  'name',
                  'parameters',
                ]),
                'title': 'tool_sum',
                'type': 'object',
              }),
              dict({
                'additionalProperties': False,
                'properties': dict({
                  'name': dict({
                    'const': 'tool_greet',
                    'description': 'Tool Name',
                    'title': 'Name',
                    'type': 'string',
                  }),
                  'parameters': dict({
                    'additionalProperties': True,
                    'description': 'Tool Parameters',
                    'properties': dict({
                      'name': dict({
                        'title': 'Name',
                        'type': 'integer',
                      }),
                    }),
                    'required': list([
                      'name',
                    ]),
                    'title': 'tool_greet',
                    'type': 'object',
                  }),
                }),
                'required': list([
                  'name',
                  'parameters',
                ]),
                'title': 'tool_greet',
                'type': 'object',
              }),
            ]),
          }),
          'title': 'Item',
          'type': 'array',
        }),
      }),
      'required': list([
        'item',
      ]),
      'title': 'AvailableTools',
      'type': 'object',
    }),
  })
# ---
